File: Greedy Algorithms/ClassPhotos.py
# Note: In a real interview it would be important to ask the interviewer if you can mutate the input array or not. This would ultimately impact the Space Complexity.

# O(nlog(n) Time | O(1) Space - Where n is the number of students
def classPhotos(redShirtHeights, blueShirtHeights):
    # Write your code here.
	redShirtHeights.sort()
	blueShirtHeights.sort()
	
	if redShirtHeights[0] > blueShirtHeights[0]:
		tallRow = redShirtHeights[:]
		shortRow = blueShirtHeights[:]
	else:
		shortRow = redShirtHeights[:]
		tallRow = blueShirtHeights[:]
	
	for i in range(0, len(redShirtHeights)):
		if tallRow[i] <= shortRow[i]:
			return False
	
    return True

File: Greedy Algorithms/ClassPhotos.java
// Solution 2: a bit more similar to the algo expert solution:
import java.util.*;

// O(nlogn) Time | O(1) space - where n is the number of students
class Program {
  public boolean classPhotos(
      ArrayList<Integer> redShirtHeights, ArrayList<Integer> blueShirtHeights) {
    Collections.sort(redShirtHeights);
		Collections.sort(blueShirtHeights);
		
		boolean redIsTall = redShirtHeights.get(0) > blueShirtHeights.get(0) ? true : false;
		
		for (int idx = 0; idx < redShirtHeights.size(); idx++) {
			if ( redIsTall) {
				if (redShirtHeights.get(idx) <= blueShirtHeights.get(idx))
					return false;
			} else {
				if (blueShirtHeights.get(idx) <= redShirtHeights.get(idx))
					return false;
				}
		}
		
		return true;
  }
}




// Solution 1: My more intuitive solution
// import java.util.*;

class Program {

  public boolean classPhotos(
      ArrayList<Integer> redShirtHeights, ArrayList<Integer> blueShirtHeights) {
    Collections.sort(redShirtHeights);
		Collections.sort(blueShirtHeights);
		ArrayList<Integer> tallRow;
		ArrayList<Integer> shortRow;
			
		if (redShirtHeights.get(0) > blueShirtHeights.get(0)) {
			tallRow = redShirtHeights;
			shortRow = blueShirtHeights;
		}
		else {
			shortRow = redShirtHeights;
			tallRow = blueShirtHeights;
		}
		
		for (int idx = 0; idx < redShirtHeights.size(); idx++) {
			if (shortRow.get(idx) >= tallRow.get(idx))
				return false;
		}
		
		return true;
  }
}

